src/librustc_borrowck/borrowck/mir/elaborate_drops.rs

```diff
@@ -0,0 +1,1094 @@
+        let mut data = InitializationData {
+            live: self.flow_inits.sets().on_entry_set_for(loc.block.index())
+                .to_owned(),
+            dead: self.flow_uninits.sets().on_entry_set_for(loc.block.index())
+                .to_owned(),
+        };
+        for stmt in 0..loc.index {
+            data.apply_location(self.ctxt.0, self.ctxt.1, &self.ctxt.2,
+                                Location { block: loc.block, index: stmt });
+        }
+        data
+    }
+
+    fn create_drop_flag(&mut self, index: MovePathIndex) {
+        let tcx = self.tcx();
+        let patch = &mut self.patch;
+        self.drop_flags.entry(index).or_insert_with(|| {
+            patch.new_temp(tcx.types.bool)
+        });
+    }
+
+    fn drop_flag(&mut self, index: MovePathIndex) -> Option<Lvalue<'tcx>> {
+        self.drop_flags.get(&index).map(|t| Lvalue::Temp(*t))
+    }
+
+    fn elaborate(mut self) -> (MirPatch<'tcx>, FnvHashMap<MovePathIndex, u32>) {
+        self.collect_drop_flags();
+        self.elaborate_drops();
+
+        self.drop_flags_on_init();
+        self.drop_flags_for_fn_rets();
+        self.drop_flags_for_args();
+        self.drop_flags_for_locs();
+
+        (self.patch, self.drop_flags)
+    }
+
+    fn path_needs_drop(&self, path: MovePathIndex) -> bool
+    {
+        match self.move_data().move_paths[path].content {
+            MovePathContent::Lvalue(ref lvalue) => {
+                let ty = self.mir().lvalue_ty(self.tcx(), lvalue)
+                    .to_ty(self.tcx());
+                debug!("path_needs_drop({:?}, {:?} : {:?})", path, lvalue, ty);
+
+                self.tcx().type_needs_drop_given_env(ty, &self.param_env)
+            }
+            _ => false
+        }
+    }
+
+    fn lvalue_concrete_strict_parent<'a>(&self, lv: &'a Lvalue<'tcx>)
+                                         -> Option<&'a Lvalue<'tcx>>
+    {
+        if let &Lvalue::Projection(ref data) = lv {
+            self.lvalue_concrete_parent(&data.base)
+        } else {
+            None
+        }
+    }
+
+    fn lvalue_concrete_parent<'a>(&self, lv: &'a Lvalue<'tcx>)
+                                  -> Option<&'a Lvalue<'tcx>>
+    {
+        let ty = self.mir().lvalue_ty(self.tcx(), lv)
+            .to_ty(self.tcx());
+        match ty.sty {
+            ty::TyArray(..) | ty::TySlice(..) | ty::TyRef(..) | ty::TyRawPtr(..) => {
+                Some(lv)
+            }
+            _ => match lv {
+                &Lvalue::Projection(ref data) => {
+                    self.lvalue_concrete_parent(&data.base)
+                }
+                _ => None
+            }
+        }
+    }
+
+    fn collect_drop_flags(&mut self)
+    {
+        for bb in self.mir().all_basic_blocks() {
+            let data = self.mir().basic_block_data(bb);
+            let terminator = data.terminator();
+            let location = match terminator.kind {
+                TerminatorKind::Drop { ref location, .. } => location,
```

Comment:
> I don't think adding the unwinds in here changes the equation significantly, but I could be wrong (in any case, with -C panic=abort, we probably ought to omit the unwinds). I imagine the unwinds don't matter because -- as you pointed out earlier -- c will be considered definitely initialized on the unwind path from each DropAndReplace.

ah, well, if `vec![]` were to unwind, that'd be different, but then we could make some other example where that's not possible.